File: BenchmarkingPackage/android/src/shared/java/com/benchmarkingpackage/PerformanceLoggerStorage.java
package com.benchmarkingpackage;

import java.util.ArrayList;
import java.util.HashMap;

public class PerformanceLoggerStorage {

    final String PAINT_START_TIME = "PAINT_START_TIME";
    final String PAINT_END_TIME = "PAINT_END_TIME";

    static HashMap<String, ArrayList<HashMap<String, Double>>> logs = new HashMap<>();
    void logStartTime(String tagName, String timestamp) {
        if (!logs.containsKey(tagName)) {
            HashMap<String, Double> map = new HashMap<>();
            ArrayList<HashMap<String, Double>> list = new ArrayList<>();
            map.put(PAINT_START_TIME, Double.valueOf(timestamp));
            list.add(map);
            logs.put(tagName, list);
        } else {
            ArrayList<HashMap<String, Double>> list = logs.get(tagName);
            HashMap<String, Double> map = new HashMap<>();
            map.put(PAINT_START_TIME, Double.valueOf(timestamp));
            assert list != null;
            list.add(map);
        }
    }

    void logEndTime(String tagName, String timestamp) {
        if (logs.containsKey(tagName)) {
            ArrayList<HashMap<String, Double>> list = logs.get(tagName);
            assert list != null;
            HashMap<String, Double> map = list.get(list.size() - 1);
            map.put(PAINT_END_TIME, Double.valueOf(timestamp));
        }
    }

    void resetLogs() {
        logs = new HashMap<>();
    }

    HashMap<String, ArrayList<HashMap<String, Double>>> getLogs() {
        return logs;
    }
}

File: BenchmarkingPackage/ios/PerformanceLoggerStorage.m
#import "PerformanceLoggerStorage.h"

@implementation PerformanceLoggerStorage

static NSMutableDictionary<NSString *, NSMutableArray<NSDictionary<NSString *, NSNumber *> *> *> *_logs = nil;

//+ (instancetype)sharedInstance {
//    static PerformanceLoggerStorage *sharedInstance = nil;
//    static dispatch_once_t onceToken;
//    dispatch_once(&onceToken, ^{
//        sharedInstance = [[self alloc] init];
//    });
//    return sharedInstance;
//}

- (instancetype)init {
    self = [super init];
    if (self) {
        if (!_logs) {
            _logs = [NSMutableDictionary dictionary];
        }
    }
    return self;
}

- (void)logStartTime:(NSString *)tagName timestamp:(NSString *)timestamp {
    if (!_logs[tagName]) {
        _logs[tagName] = [NSMutableArray arrayWithObject:@{@"PAINT_START_TIME": @([timestamp doubleValue])}];
    } else {
        [_logs[tagName] addObject:@{@"PAINT_START_TIME": @([timestamp doubleValue])}];
    }
}

- (void)logEndTime:(NSString *)tagName timestamp:(NSString *)timestamp {
    NSMutableArray<NSDictionary<NSString *, NSNumber *> *> *list = _logs[tagName];
    
    if (list && list.count > 0) {
        NSMutableDictionary<NSString *, NSNumber *> *lastEntry = [list lastObject].mutableCopy;
        lastEntry[@"PAINT_END_TIME"] = @([timestamp doubleValue]);

        // Replace the last entry in the array with the modified one
        [list replaceObjectAtIndex:list.count - 1 withObject:lastEntry];
    }
}
- (void)resetLogs {
    [_logs removeAllObjects];
}

- (NSDictionary<NSString *, NSArray<NSDictionary<NSString *, NSNumber *> *> *> *)getLogs {
    return [_logs copy];
}

@end
